### DISCUSSION TITLE
Use password variables immediately if filled...

### DISCUSSION THREAD

**Discussion by ghost** [Waboodoo/HTTP-Shortcuts#316] [Ideas]
Hello,

I use this nice&convenient app (with well targeted scripting) for some home automation (Domoticz based, using it's http json API).

I have a suggestion for password type variables handling: An option may be added, if some password was already stored, to use it immediately and only show the password window in case of fail (to be able to update value immediately in case of server side change, for instance).

This would make remote commands needing login smoother IMO. Default may be kept as current, to avoid disturbing users that currently takes this as a confirmation step... and not already using shortcut confirmation option!

Thanks&Regards.

**Comment by Waboodoo** [comment]
Thanks for this input.
There is the problem though that the password variable itself has no concept of success or failure, all it does is provide a value. The failure happens only much later, and it is highly use case specific what the failure looks like, so I don't see an easy way of accomplishing this.

I have an alternative idea though, which would achieve the same thing: In the next version (2.26.0) I'm introducing a new Scripting function `promptPassword()` which can be used to prompt for a password. This way, instead of a password variable, you could use a regular static variable to store your password, and in case of a failure you could have your shortcut run a code snippet that queries for a new password using this function, stores the new password in that static variable and then (if need be) rerun the shortcut.

So in your shortcut's "Run on Failure" text body, on the Scripting screen, you could have code that looks like this (assuming you have a static variable named "myPassword" which stores the password):
```js
if (response && response.statusCode == 401) { // you might need to adapt this check, depending on what "failure" means in your case
    const oldPassword = getVariable('myPassword');
    const newPassword = promptPassword('Please enter your password', oldPassword);
    setVariable('myPassword', newPassword);
    enqueueShortcut(''); // this will re-run the current shortcut
    abort(); // this will prevent the displaying of an error dialog
}
```

**Reply by ghost** [reply]
Hello,
Thanks for your answer & proposition, I understand this would not be feasible (+may break other users use-cases): This looks good & more generic. The acronym for this must be... KISS, in fact!
Regards.

**Comment by ghost** [comment]
Hello,
Can confirm your suggestion does the expected job smoothly with latest version. Combined with settings lock, just to avoid someone getting hands on your unlocked phone being able to get this sensitive data too easily, directly from normal static variable or generating a settings export file.
Thanks!
